aten/src/ATen/function_wrapper.py

```diff
@@ -222,16 +253,77 @@ def TypedDict(name, attrs, total=True):  # type: ignore
 }
 """)
 
+COLLAPSED_FACTORY_DEFINITION = CodeTemplate("""\
+inline ${return_type} ${api_name}(${collapsed_formals}) { 
+    return _${api_name}(${expanded_native_actuals});
+}
+""")
+
+COLLAPSED_METHOD_DEFINITION = CodeTemplate("""\
+inline ${return_type} Tensor::${api_name}(${collapsed_formals}) const { 
+    c10::optional<ScalarType> dtype = c10::nullopt;
+    c10::optional<Layout> layout = c10::nullopt;
+    c10::optional<Device> device = c10::nullopt;
+    c10::optional<bool> pin_memory = c10::nullopt;
+
+    if (options.dtype_opt().has_value()) {
+        dtype = typeMetaToScalarType(options.dtype());
+    }
+
+    if (options.layout_opt().has_value()) {
+        layout = options.layout();
+    }
+    
+    if (options.device_opt().has_value()) {
+        device = options.device();
+    }
+
+    if (options.pinned_memory_opt().has_value()) {
+        pin_memory = options.pinned_memory();
+    }
+
+    return _${api_name}(${expanded_native_actuals});
+}
+""")
+
+# This is a special case for '.to' operator. It should be cleaned up
+# issue #30405
+COLLAPSED_METHOD_TO_DEFINITION = CodeTemplate("""\
+inline ${return_type} Tensor::${api_name}(${collapsed_formals}) const { 
+    TORCH_CHECK(options.requires_grad_opt() == c10::nullopt,
+         "to(options) expects unset requires_grad flag, but got "
+         "options.requires_grad set as ", options.requires_grad());
```

Comment:
Is this the only reason it's special cased? In fact, this raises a broader question, which is, why isn't `requires_grad` part of the expanded arguments.

Reply:
It... sure looks like it's part of TensorOptions to me!

```
  struct C10_API TensorOptions {

...

    caffe2::TypeMeta dtype_ = caffe2::TypeMeta::Make<float>(); // 64-bit
    Device device_ = at::kCPU; // 32-bit
    Layout layout_ = at::kStrided; // 8-bit
  
    // Bitmask required here to get this to fit inside 32 bits (or even 64 bits,
    // for that matter)
  
    bool requires_grad_     : 1;
    bool pinned_memory_     : 1;
  
  
    bool has_device_        : 1;
    bool has_dtype_         : 1;
    bool has_layout_        : 1;
    bool has_requires_grad_ : 1;
    bool has_pinned_memory_ : 1;
```